onnxruntime/core/framework/tensorprotoutils.cc

```diff
@@ -726,9 +748,30 @@ Status TensorProtoToMLValue(const Env& env, const ORTCHAR_T* model_path,
   // Note: We permit an empty tensor_shape_vec, and treat it as a scalar (a tensor of size 1).
   TensorShape tensor_shape{tensor_shape_vec};
 
+  std::vector<std::function<void(void)>> tensor_deleters;
+  ResourceGuard exitguard([&tensor_deleters]() {
+    for (const auto& d : tensor_deleters) {
```

Comment:
What's the use case where we need this sort of automated cleanup?

Reply:
Three cases:
1. when the tensor raw data comes from a separate file, in this case we have an extra buffer to free.
2. when the tensor is string type, we need to release all strings before the buffer is released.
3. when the pre-allocated MemBuffer is ownership carrying, need to make sure it is released.

And this scope guard is there to make sure we release all buffers during all error cases.